# CÂU 1: PHÂN TÍCH HÀNH VI MUA LẠI CỦA KHÁCH HÀNG

**1. Yêu cầu bài toán:**
Tính trung vị (median) số ngày giữa hai lần mua hàng liên tiếp đối với những khách hàng có từ 2 đơn hàng trở lên. (Sử dụng file `orders.csv`).

**2. Phương pháp tiếp cận & Suy luận:**
Để tính được khoảng cách giữa 2 lần mua, ta cần thực hiện các bước logic sau:
* **Bước 1:** Chuẩn bị dữ liệu. Chuyển đổi cột ngày tháng `order_date` sang định dạng Datetime chuẩn để có thể làm phép trừ toán học.
* **Bước 2:** Nhóm và Sắp xếp. Dữ liệu ban đầu đang nằm lộn xộn. Ta cần gom các đơn hàng của cùng một người (`customer_id`) lại với nhau, sau đó sắp xếp theo thứ tự thời gian từ cũ nhất đến mới nhất.
* **Bước 3:** Tính khoảng cách. Dùng hàm `diff()` để lấy ngày của đơn hàng sau trừ đi ngày của đơn hàng liền trước nó.
* **Bước 4:** Lọc và Tính trung vị. Những khách hàng chỉ mua 1 lần (hoặc đơn hàng đầu tiên của mỗi khách) sẽ không có khoảng cách, hàm sẽ trả về rỗng (NaN). Ta sẽ tính trung vị (median) trên các giá trị hợp lệ còn lại.

In [3]:
# Import thư viện cần thiết
import pandas as pd
from google.colab import drive

# Kết nối với Google Drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


**Thực hiện Bước 1 & Bước 2: Tải dữ liệu và Tiền xử lý**
Ta sẽ load file `orders.csv`, ép kiểu dữ liệu thời gian và tiến hành sắp xếp (sort). Việc sắp xếp là bắt buộc để hàm `diff()` ở bước sau trừ đúng đơn hàng liền kề của cùng một khách hàng.

In [4]:
# Đọc dữ liệu từ Drive
file_path = '/content/drive/MyDrive/datathon-2026-round-1/orders.csv'
df_orders = pd.read_csv(file_path)

# Ép kiểu cột order_date về định dạng datetime
df_orders['order_date'] = pd.to_datetime(df_orders['order_date'])

# Sắp xếp dữ liệu ưu tiên theo customer_id, sau đó tới order_date tăng dần
df_orders = df_orders.sort_values(by=['customer_id', 'order_date'])

# Hiển thị thử 5 dòng đầu tiên sau khi sắp xếp
df_orders.head()

,order_id,order_date,customer_id,zip,order_status,payment_method,device_type,order_source
4023,5280,2012-07-25,1,15201,delivered,cod,desktop,paid_search
143252,184922,2014-05-31,1,15201,returned,credit_card,mobile,referral
238890,308113,2015-07-31,1,15201,delivered,cod,mobile,paid_search
374571,483190,2017-04-23,1,15201,delivered,cod,mobile,paid_search
544446,702081,2020-02-24,1,15201,delivered,credit_card,mobile,organic_search


**Thực hiện Bước 3 & Bước 4: Tính toán và Đưa ra kết luận**
Bây giờ, ta nhóm dữ liệu theo từng khách hàng (`groupby('customer_id')`) và tính độ lệch ngày (`diff()`). Sau khi có cột độ lệch, áp dụng hàm `.median()` để tìm ra giá trị trung vị theo đúng yêu cầu đề bài.

In [5]:
# Tính số ngày giữa 2 lần mua liên tiếp
df_orders['inter_order_gap'] = df_orders.groupby('customer_id')['order_date'].diff().dt.days

# Tính giá trị trung vị (tự động bỏ qua các giá trị NaN của đơn hàng đầu tiên)
median_gap = df_orders['inter_order_gap'].median()

print(f"--> Trung vị số ngày giữa hai lần mua liên tiếp là: {median_gap} ngày.")

--> Trung vị số ngày giữa hai lần mua liên tiếp là: 144.0 ngày.


**KẾT LUẬN CUỐI CÙNG:**
Kết quả tính toán thực tế cho ra trung vị là **144 ngày**.

Đối chiếu với các phương án của đề bài:

A) 30 ngày

B) 90 ngày

C) 144 ngày

D) 365 ngày


**=> Đáp án được chọn là: C**